src/main/java/com/buglife/sdk/BugReporterImpl.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (C) 2017 Buglife, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.buglife.sdk;
+
+import android.app.job.JobInfo;
+import android.app.job.JobScheduler;
+import android.content.Context;
+import android.os.Build;
+import android.os.PersistableBundle;
+import android.support.annotation.RequiresApi;
+
+import com.buglife.sdk.reporting.BugReporter;
+import com.buglife.sdk.reporting.SubmitReportLegacyService;
+import com.buglife.sdk.reporting.SubmitReportService;
+
+import org.json.JSONException;
+
+final class BugReporterImpl implements BugReporter {
+    private final Context mContext;
+
+    BugReporterImpl(Context context) {
+        mContext = context;
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+            // Kick off submit report service in case there are cached reports still
+            // available.
+            SubmitReportLegacyService.start(mContext);
+        }
+    }
+
+    @Override public void report(Report report) {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+            reportWithJobScheduler(report);
+        } else {
+            reportWithLegacy(report);
+        }
+    }
+
+    private void reportWithLegacy(Report report) {
+        try {
+            String jsonReport = report.toJSON().toString();
+            SubmitReportLegacyService.start(mContext, jsonReport);
+        } catch (JSONException e) {
+            Log.e("Failed to encode report!", e);
```

Comment:
Can we bubble up this error somehow?

Reply:
@schukin Do you mean surfacing this error to the UI? We could display a Toast or re-throw the exception. My initial thought would be to re-throw since the having a badly formatted report would be a pretty serious error.

Reply:
Yeah, we should surface this error to the UI. That way when users encounter it, they at least have a vague clue that something went wrong and can help us figure it out. I would think it'd make sense to re-throw the exception at this level. Up the chain, I think we'd either want to throw a RuntimeException & let the app crash, or show a Toast so the user knows something went wrong. Thoughts?

Reply:
@schukin would you be opposed to doing both, throwing the exception and showing a toast here?

Reply:
@alexfu that sounds reasonable 👍